packages/dataflow/src/__tests__/readme.ts

```diff
@@ -1,25 +1,78 @@
+            filter(n => n <= 0.5),              //  Filter out numbers > 0.5  
+            sensor(c2),                         //  Keep running count of filtered rows
+            map((n, idx) =>                     //  Convert to JSON Object 
+                ({ index: idx, value: n })),
+            filter(row => row.index % 2 === 0), //  Filter even row indecies 
+            sensor(c3),                         //  Keep running count of final rows
+            sensor(m1),                         //  Track largest value
+            first(3)                            //  Take first 3 rows
+        );
 
-    //  Take first 3 rows
-    first(3)
-);
-console.log(...outIterable);
+        console.log(`Counts: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}`);
+        // [1] => Counts: undefined, undefined, undefined
+        const outIterable = p1(generate(Math.random, 1000));
+        console.log(`Counts: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}`);
+        // [2] => Counts: undefined, undefined, undefined
+        console.log(JSON.stringify([...outIterable]));
+        // [3] => [{"index":0,"value":0.19075931906641008},{"index":2,"value":0.4873469062925415},{"index":4,"value":0.4412516774100035}]
+        console.log(`Counts: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}, ${m1.peek()}`);
+        // [4] => Counts: 6, 5, 3, 0.4873469062925415
+
+        const outArray = [...p1([0.7, 0.5, 0.4, 0.8, 0.3, 1])];
+        console.log(JSON.stringify(outArray));
+        // [5] => [{"index":0,"value":0.5},{"index":2,"value":0.3}]
+        console.log(`Counts: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}, ${m1.peek()}`);
+        // [6] => Counts:  6, 3, 2, 0.5
+
+        expect(outArray.length).to.equal(2);
+
+        for (const row of p1(generate(Math.random, 1000000))) {
+            console.log(`${row.index}: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}, ${m1.peek()}`);
+        }
+    });
+
+    it("interesting example", () => {
+        const c1 = count();
+        const c2 = count();
+        const c3 = count();
+        const m1 = max(row => row.value);
+
+        const p1 = pipe(
+            sensor(c1),                         //  Keep running count of input
+            filter(n => n <= 0.5),              //  Filter out numbers > 0.5  
+            sensor(c2),                         //  Keep running count of filtered rows
+            map((n, idx) =>                     //  Convert to JSON Object 
+                ({ index: idx, value: n })),
+            filter(row => row.index % 2 === 0), //  Filter even row indecies 
+            sensor(c3),                         //  Keep running count of final rows
+            sensor(m1),                         //  Track largest value
+        );
+
+        for (const row of p1(generate(Math.random, 1000000))) {
+            if (row.index % 100000 === 0) {
+                console.log(`${row.index}: ${c1.peek()}, ${c2.peek()}, ${c3.peek()}, ${m1.peek()}`);
+            }
+        }
```

Comment:
This _is_ interesting. So each row yielded by the generator is piped through p1 one at a time? I wonder if the sensors prevent garbage collection?

Reply:
In this example there is no memory overhead as its one row at a time "flowing" through the pipe, so no need for garbage collection at all...

Reply:
Unfortunatly the same isn't true for all activities (I am looking at you `sort`).